presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java

```diff
@@ -558,6 +564,31 @@ public void beginQuery(Session session, Set<ConnectorId> connectors)
             ConnectorMetadata metadata = getMetadata(session, connectorId);
             ConnectorSession connectorSession = session.toConnectorSession(connectorId);
             metadata.beginQuery(connectorSession);
+            registerCatalogForQueryId(session.getQueryId(), metadata);
+        }
+    }
+
+    private void registerCatalogForQueryId(QueryId queryId, ConnectorMetadata metadata)
+    {
+        catalogsByQueryId.putIfAbsent(queryId.getId(), new ArrayList<>());
+        catalogsByQueryId.get(queryId.getId()).add(metadata);
+    }
+
+    @Override
+    public void cleanupQuery(Session session)
+    {
+        try {
+            Collection<ConnectorMetadata> catalogs = catalogsByQueryId.get(session.getQueryId().getId());
```

Comment:
When you have two catalogs of the same type and both participate in query, then you want both catalogs being notified.
E.g.:
`INSERT INTO hive_finance.reports.stats SELECT * FROM hive_hr.salaries.engineers;`
Both: `hive_hr` and `hive_finance` should be notified.